WIP: Thredds download function

bailsofhay opened the pull request:
This function is the start of a general function that should be able to download data from THREDDS.

## Description
The user can download any THREDDS datasets as long as they have site information, dates, and the base URLS needed to grab the datasets. Base URLS need the dataset name, but not the dates or locations. There will likely be issues pulling data sets in the future since they are not all stored in similarly structured directory formats like AVHRR LAI (which this function is based on). Updates to the function will occur as we start playing with more datasets from THREDDS.

## Motivation and Context
We need an easy way to grab more datasets and THREDDS has many already, and is a great place to store future datasets that we can use.

## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [ ] Immediately
- [ ] Within one week
- [X] When possible
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

Commit:
added updated thredds function

bailsofhay commented:
@ankurdesai I had to remove the earlier PR I made last week with the SDA file changes attached. this is the new PR for the same Thredds function, but without non-related file changes.

bailsofhay commented:
Turns out there was already a branch named thredds that was messing up my pr....... :/

infotroph commented on modules/data.remote/R/download.thredds.R:
```diff
@@ -20,84 +30,124 @@
 ##'
-download.thredds.AGB <- function(outdir = NULL, site_ids, run_parallel = FALSE, 
-                                    ncores = NULL) {
+download.thredds.data <- function(outdir = NULL, site_info, dates = c("19950201", "19961215"), 
```
I recommend *not* providing a default value for `dates`, so that it fails immediately and clearly when users don't provide them
```suggestion
download.thredds.data <- function(outdir = NULL, site_info, dates, 
```

infotroph replied:
And while I'm thinking about it, consider moving `outdir` to come after  later in the argument list -- it's best to arrange arguments such that everything without a default value comes before everything with a default value, even if the default is NULL.

infotroph commented on modules/data.remote/R/download.thredds.R:
```diff
@@ -20,84 +30,124 @@
+    result <- RCurl::getURL(paste(dir_url, "catalog.html", sep = "/"), verbose=F,ftp.use.epsv=TRUE, dirlistonly = TRUE)
+    files = XML::getHTMLLinks(result)
+    
+    date_year_range = unique(range(c(year(as.Date(dates[1], "%Y")), year(as.Date(dates[2], "%Y")))))
```
`as.Date` on a Date object and `range` on two values are both no-ops, so this can be simplified to:
```suggestion
    date_year_range = unique(lubridate::year(dates))
```

infotroph reviewed (changes requested):
I see that this is still WIP so I hope my review isn't premature, but a few interface-level comments that are probably easier to address sooner rather than later. 

Might also be useful to add a fully updated example that can double as a test case while you're working on the details.

mdietze commented on modules/data.remote/R/download.thredds.R:
```diff
@@ -20,84 +30,124 @@
     
-    # option to save output dataset to directory for user.
-    if (!(is.null(outdir)))
+    extract_nc = function(site_info, url, run_parallel)
```
I'll also add that there's already an extract.nc function in data.atmosphere, which is going to be very confusing. Furthermore, after we clean up the function naming (i.e. finally switch . to _ ) it could easily lead to genuine conflicts. One option is to rename this function, since it's the new one, and because the args makes it clear that this function isn't just extracting it's also downloading. Alternatively, you could take a look at the other function and think about whether you could generalize it (without breaking it's current I?O and behaviors) to serve both cases.

Either way, the function definitely needs Roxygen documentation, especially since site_info is a list of stuff

infotroph replied:
Good point, @mdietze -- renaming or generalizing seems wise. But do note that this is a helper that isn't exported and probably doesn't need to be, which (1) mitigates the name conflict issues (though not the potential for confusion) and (2) personally I think it makes Roxygen docs less important -- to me it's clear in context that site_info will be the same thing here as it is in download_thredds_data.

infotroph commented:
Once merged this will close #2415, correct?

mdietze commented on modules/data.remote/R/download.thredds.R:
```diff
@@ -1,17 +1,27 @@
-##' @param outdir Where to place output
-##' @param site_ids What locations to download data at? 
+##' @param outdir file location to place output
+##' @param site_info information about the site. i.e. site_id, latitude, longitude 
```
site_info is insufficiently documented. Are there arguments other than site_id, latitude, and longitude? Is site_id the one in BETY or something else? are lat/lon decimal degrees or something else? Is this supposed to be in a list, or a vector, or a data frame, or something else? Are multiple sites allowed

bailsofhay replied:
ahhh. I see what you mean. I'll update this to be more informative

mdietze commented on modules/data.remote/R/download.thredds.R:
```diff
@@ -1,17 +1,27 @@
-##' @title download.thredds.AGB
-##' @name  download.thredds.AGB
+##' @title download.thredds.data
+##' @name  download.thredds.data
```
Is `.data` really needed at the end of this name?

bailsofhay replied:
fixed

bailsofhay commented:
@infotroph, @ashiklom or @mdietze. What is the standard way to code in warnings. For example, I want a warning to pop up and kill data extraction if the user puts in a date range that the data doesn't exist during.

mdietze commented:
Warning, errors, and logs should be handled through the logger functions in the PEcAn.utils package

Commit:
some updated changes

Commit:
added outdir option in function

Commit:
added date corrections

Commit:
updated @params

Commit:
updated date_year_range

Commit:
separated nc extract function from download function

Commit:
fixed documentation issue

Commit:
some other changes i dont remember

Commit:
thredds function updates

Commit:
some changes I dont remember

Commit:
update thredds function

infotroph committed:
Merge branch 'develop' into thredds_function

infotroph committed:
roxygen

infotroph committed:
quick cleanup of package check results

* delete duplicated identical definitions of get_site_info and extract_thredds_nc
	(multiple conflicting definitions remain of download_thredds and extract_thredds_nc)
* clean up long lines in @examples
* Roxygen format cleanup
* convert require() to pkg::

infotroph committed:
avoid conflicting definitions of same name

infotroph committed:
avoid using RCurl

infotroph committed:
inconsistent param names

infotroph committed:
fix "no visible binding" check notes

infotroph committed:
run styler

mdietze committed:
Merge branch 'develop' into thredds_function
